Source/FieldSolver/SpectralSolver/SpectralFieldDataHankel.cpp

```diff
@@ -0,0 +1,440 @@
+
+    // Loop over boxes and allocate the corresponding plan
+    // for each box owned by the local MPI proc.
+    for (amrex::MFIter mfi(spectralspace_ba, dm); mfi.isValid(); ++mfi){
+        amrex::IntVect grid_size = realspace_ba[mfi].length();
+#ifdef AMREX_USE_GPU
+        // Create cuFFT plan.
+        // This is alway complex to complex.
+        // This plan is for one azimuthal mode only.
+        cufftResult result;
+        int fft_length[] = {grid_size[1]};
+        int inembed[] = {grid_size[1]};
+        int istride = grid_size[0];
+        int idist = 1;
+        int onembed[] = {grid_size[1]};
+        int ostride = grid_size[0];
+        int odist = 1;
+        int batch = grid_size[0]; // number of ffts
+        result = cufftPlanMany(&forward_plan[mfi], 1, fft_length, inembed, istride, idist,
+                               onembed, ostride, odist, CUFFT_Z2Z, batch);
+        if (result != CUFFT_SUCCESS) {
+           amrex::Print() << " cufftPlanMany failed! \n";
+        }
+        // The backward plane is the same as the forward since the direction is passed when executed.
+
+#else
+        // Create FFTW plans.
+        fftw_iodim dims[1];
+        fftw_iodim howmany_dims[2];
+        dims[0].n = grid_size[1];
+        dims[0].is = grid_size[0];
+        dims[0].os = grid_size[0];
+        howmany_dims[0].n = n_rz_azimuthal_modes;
+        howmany_dims[0].is = grid_size[0]*grid_size[1];
+        howmany_dims[0].os = grid_size[0]*grid_size[1];
+        howmany_dims[1].n = grid_size[0];
+        howmany_dims[1].is = 1;
+        howmany_dims[1].os = 1;
+        forward_plan[mfi] =
+            // Note that AMReX FAB are Fortran-order.
+            fftw_plan_guru_dft(1, // int rank
+                               dims,
+                               2, // int howmany_rank,
+                               howmany_dims,
+                               reinterpret_cast<fftw_complex*>(tempHTransformed[mfi].dataPtr()), // fftw_complex *in
+                               reinterpret_cast<fftw_complex*>(tmpSpectralField[mfi].dataPtr()), // fftw_complex *out
+                               FFTW_FORWARD, // int sign
+                               FFTW_ESTIMATE); // unsigned flags
+        backward_plan[mfi] =
+            fftw_plan_guru_dft(1, // int rank
+                               dims,
+                               2, // int howmany_rank,
+                               howmany_dims,
+                               reinterpret_cast<fftw_complex*>(tmpSpectralField[mfi].dataPtr()), // fftw_complex *in
+                               reinterpret_cast<fftw_complex*>(tempHTransformed[mfi].dataPtr()), // fftw_complex *out
+                               FFTW_BACKWARD, // int sign
+                               FFTW_ESTIMATE); // unsigned flags
+#endif
+
+        // Create the Hankel transformer for each box.
+        std::array<amrex::Real,3> xmax = WarpX::UpperCorner(mfi.tilebox(), 0); // lev=0 is explicit
```

Comment:
Could you explain why `lev=0` is imposed here?
Could we obtain `rmax` by other means, so that it would generalize to any level (maybe by passing the radial grid size `dr` as argument to the constructor, and using the info in `realspace_ba`)?

Reply:
It would be easiest to pass lev in. Will this ever likely be an issue though? Would the radial extent of a refined level ever be smaller than the coarser level, since this depends on having a radial PML boundary?